pkg/controller/modify_volume.go

```diff
@@ -0,0 +1,135 @@
+// to Pending if VAC does not exist and proceeds to trigger ModifyVolume if VAC exists
+func (ctrl *resizeController) validateVACAndModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	// The controller only triggers ModifyVolume if pvcSpecVacName is not nil nor empty
+	pvcSpecVacName := pvc.Spec.VolumeAttributesClassName
+	// Check if pvcSpecVac is valid and exist
+	vacObj, exists, err := ctrl.volumeAttributesClasses.GetByKey(*pvcSpecVacName)
+	if exists && err == nil {
+		// Mark pvc.Status.ModifyVolumeStatus as in progress
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInProgress, nil)
+		// Record an event to indicate that external resizer is modifying this volume.
+		ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeModify,
+			fmt.Sprintf("external resizer is modifying volume %s", pvc.Name))
+		return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vacObj.(*storagev1alpha1.VolumeAttributesClass), pvcSpecVacName)
+	} else {
+		// Mark pvc.Status.ModifyVolumeStatus as pending
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumePending, nil)
+		return pvc, pv, nil, false
+	}
+}
+
+// func controllerModifyVolumeWithTarget trigger the CSI ControllerModifyVolume API call
+// and handle both success and error scenarios
+func (ctrl *resizeController) controllerModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	vacObj *storagev1alpha1.VolumeAttributesClass,
+	pvcSpecVacName *string) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	var err error
+	pvc, pv, err = ctrl.callModifyVolumeOnPlugin(pvc, pv, vacObj)
+	if err == nil {
+		klog.V(4).Infof("Update volumeAttributesClass of PV %q to %s succeeded", pv.Name, *pvcSpecVacName)
+		return pvc, pv, nil, true
+	} else {
+		if status, ok := status.FromError(err); ok && status.Code() == codes.InvalidArgument {
+			// Mark pvc.Status.ModifyVolumeStatus as infeasible
+			pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInfeasible, err)
+		} else {
+			ctrl.updateConditionBasedOnError(pvc, err)
+			if !isFinalError(err) {
+				// update conditions and cache pvc as uncertain
+				ctrl.uncertainPVCs.Add(pvc)
+			} else {
+				ctrl.removePVCFromModifyVolumeUncertainCache(pvc)
+			}
+		}
+		// Record an event to indicate that resize operation is failed.
+		ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeModifyFailed, err.Error())
+		return pvc, pv, err, true
+	}
+}
+
+func (ctrl *resizeController) callModifyVolumeOnPlugin(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	vac *storagev1alpha1.VolumeAttributesClass) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error) {
+	err := ctrl.resizer.Modify(pv, vac.Parameters)
+
+	if err != nil && isFinalError(err) {
```

Comment:
So if the error is not nil but not a final error, then modifyVolume is considered as completed? Seems incorrect..

Reply:
Error should be returned and handled in the controllerModifyVolumeWithTarget, let me update that.